### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by huonglarne**
Refactor the previous function for more elegant code

### CODE REVIEW

File: trl/trainer/moreh_utils.py [hunk 1]
```diff
@@ -11,17 +11,13 @@
...
     start_time = time.time()
-    epoch_loss_list = []
+    epoch_start = 0
 
     def on_step_begin(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
         # count the time after the logging step
         if state.global_step == 0 or state.global_step % args.logging_steps == 1:
             self.start_time = time.time()
 
-    def on_step_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
```
Comment (yeomni) [inline trl/trainer/moreh_utils.py:21]: Just out of curiosity, does this mean we won't use average train loss when we use `steps` for logging strategy?

File: trl/trainer/moreh_utils.py [hunk 2]
```diff
@@ -37,14 +33,17 @@ def on_log(self, args: TrainingArguments, state: TrainerState, control: TrainerC
...
-        epoch_loss = sum(self.epoch_loss_list) / len(self.epoch_loss_list)
-        mlflow.log_metric("epoch_loss", epoch_loss, step=state.epoch)
-        self.epoch_loss_list = []
+        all_loss_of_this_epoch = []
+        for each in state.log_history[self.epoch_start:]:
+            if 'loss' in each:
+                all_loss_of_this_epoch.append(each['loss'])
+
+        epoch_loss = sum(all_loss_of_this_epoch) / len(all_loss_of_this_epoch)
+        mlflow.log_metric("epoch_loss", epoch_loss, step=int(state.epoch))
```
Comment (yeomni) [inline trl/trainer/moreh_utils.py:44]: How can it be treated differently? Its step number is epoch-based while `on_log` takes global_step for step.

Comment (huonglarne) [reply trl/trainer/moreh_utils.py:44]: @yeomni the "step" arg in mlflow isnt the same as step in training, it's the number on the x-axis signifying the space between values. Since here I'm logging to mlflow after each epoch, the step is the same as epoch number

File: trl/trainer/moreh_utils.py (hunk 1 above)
Comment (huonglarne) [reply trl/trainer/moreh_utils.py:21]: I don't understand the question, can you explain more?

Comment (yeomni) [reply trl/trainer/moreh_utils.py:21]: I'm wondering why this `on_step_end` method was deleted in the latest commit. I just wanted to double check that you won't need this method since we are going to measure the average training loss only when the logging_strategy is set `epoch`, rather than `steps`.

Comment (huonglarne) [reply trl/trainer/moreh_utils.py:21]: @yeomni ah I see. In the old code, I use on_step_end to collect the loss after each step but in the new code, I collect the loss directly from the state_history so this method is redundant.
The logging method is still steps, not epoch. I add a new function to collect the epoch log in addition to each step's logging.

**Review by yeomni — APPROVED** [review]
LGTM
